feat: Use the Serde attributes and check for conflicting attributes

Kriskras99 opened the pull request:
This commit changes the attribute parsing in `#[derive(AvroSchema)]` to also use the Serde attributes. It will use the Serde attributes and check that they match the Avro attributes if provided. It will also check that known incompatible Serde attributes result in a compile error instead of a runtime panic.

Testing for the compile errors is done using `trybuild`. These tests run only on nightly as the output can vary between compiler versions (otherwise MSRV and latest stable could have different outputs and break the entire CI).

For some reason Github closed #373, not sure why. This is the exact same commit.

martin-g commented on avro_derive/src/attributes/serde.rs:
```diff
@@ -0,0 +1,298 @@
+    /// This does something with tags, which are incompatible.
+    #[darling(default)]
+    pub variant_identifier: bool,
+    /// This does something with tags,  `` which are incompatible.
```
What is the purpose of the empty backticks here ?

martin-g commented on avro_derive/src/attributes/mod.rs:
```diff
@@ -0,0 +1,202 @@
+        if serde.transparent {
+            errors.push(syn::Error::new(
+                span,
+                "AvroSchema derive does not support Serde transparent",
```
```suggestion
                "AvroSchema derive does not support Serde `transparent` attribute",
```

martin-g commented on avro_derive/src/attributes/mod.rs:
```diff
@@ -0,0 +1,202 @@
+        if avro.skip && !serde.skip {
+            errors.push(syn::Error::new(
+                span,
+                "`#[avro(skip)]` requires `#[serde(skip)]`, it's also deprecated. Please use only `#[serde(skip)]`"
```
Is there a way to report the usage of a deprecated `avro(attribute)` as a warning without depending on the value of its respective `serde` counter-part ?
I.e. pseudo code:
```
if avro.skip { report a warning like "#[avro(skip)] is deprecated. Use #[serde(skip)] instead" }

if avro.skip != serde.skip {report an error, without the ", it's also deprecated" text}
```

Currently the user will be notified that a deprecated avro attribute is used only if the avro value differs from the serde's one.

martin-g replied:
Sounds good!
Let's do it!
Then once it is in `stable` and our MSRV is at least this stable version we could delete just this attribute to enable it for all.

martin-g commented on avro_derive/src/attributes/mod.rs:
```diff
@@ -0,0 +1,202 @@
+        {
+            errors.push(syn::Error::new(
+                span,
+                "`#[serde(skip_serializing)]` and `#[serde(skip_serializing_if)]` require `#[avro(default = \"..\")]"
```
```suggestion
                "`#[serde(skip_serializing)]` and `#[serde(skip_serializing_if)]` require `#[avro(default = \"..\")]`"
```

martin-g commented on avro_derive/src/attributes/mod.rs:
```diff
@@ -0,0 +1,202 @@
+                "`#[avro(rename = \"..\")]` must match `#[serde(rename = \"..\")]`, it's also deprecated. Please use only `#[serde(rename  = \"..\")]`"
+            ));
+        }
+        if serde.alias != avro.alias {
```
This would fail if only `#[serde(alias= ...)]` is used because `avro.alias` will be resolved to an empty Vec.
```suggestion
        if !avro.alias.is_empty() && serde.alias != avro.alias {
```

martin-g commented:
A PR has been opened at Derling repo with an example how to parse `serde`: 
https://github.com/TedDriggs/darling/pull/412/changes#diff-9df22bcc91f2a6d5114da4fff9f20353fb0171013e7a7fa385cea38e6130f691

We may borrow something useful from it.

martin-g commented on avro_derive/src/case.rs:
```diff
@@ -59,15 +72,20 @@ static RENAME_RULES: &[(&str, RenameRule)] = &[
-            }
+    pub fn from_str(rename_all_str: &str) -> Result<Self, ParseError> {
+        match rename_all_str {
+            "lowercase" => Ok(Self::LowerCase),
```
Why this no more uses `RENAME_RULES` ? Now the rules are duplicated.
```rust
          RENAME_RULES
            .iter()
            .find(|(name, _)| *name == rename_all_str)
            .map(|(_, rule)| *rule)
            .ok_or_else(|| ParseError {
                unknown: rename_all_str.to_string(),
            })
```

martin-g commented on avro_derive/src/attributes/mod.rs:
```diff
@@ -0,0 +1,206 @@
+        if avro.rename.is_some() && serde.rename != avro.rename {
+            errors.push(syn::Error::new(
+                span,
+                "`#[avro(rename = \"..\")]` must match `#[serde(rename = \"..\")]`, it's also deprecated. Please use only `#[serde(rename  = \"..\")]`"
```
```suggestion
                "`#[avro(rename = \"..\")]` must match `#[serde(rename = \"..\")]`, it's also deprecated. Please use only `#[serde(rename = \"..\")]`"
```

martin-g commented on avro_derive/tests/ui/avro_rs_226_skip_serializing.rs:
```diff
@@ -0,0 +1,11 @@
+use apache_avro::AvroSchema;
```
```suggestion
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

use apache_avro::AvroSchema;
```

martin-g commented on avro_derive/tests/ui/avro_rs_226_skip_serializing_if.rs:
```diff
@@ -0,0 +1,11 @@
+use apache_avro::AvroSchema;
```
```suggestion
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

use apache_avro::AvroSchema;
```

martin-g commented on avro_derive/tests/ui/avro_rs_373_remote.rs:
```diff
@@ -0,0 +1,16 @@
+use apache_avro::AvroSchema;
```
```suggestion
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

use apache_avro::AvroSchema;
```

martin-g commented on avro_derive/tests/ui/avro_rs_373_skip_de_serializing.rs:
```diff
@@ -0,0 +1,11 @@
+use apache_avro::AvroSchema;
```
```suggestion
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

use apache_avro::AvroSchema;
```

martin-g commented on avro_derive/tests/ui/avro_rs_373_tag_content_enum.rs:
```diff
@@ -0,0 +1,10 @@
+use apache_avro::AvroSchema;
```
```suggestion
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

use apache_avro::AvroSchema;
```

martin-g commented on avro_derive/tests/ui/avro_rs_373_transparent.rs:
```diff
@@ -0,0 +1,15 @@
+use apache_avro::AvroSchema;
```
```suggestion
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

use apache_avro::AvroSchema;
```

martin-g commented on avro_derive/src/attributes/mod.rs:
```diff
@@ -0,0 +1,206 @@
+                "`#[avro(alias = \"..\")]` must match `#[serde(alias = \"..\")]`, it's also deprecated. Please use only `#[serde(alias  = \"..\")]`"
+            ));
+        }
+        if serde.skip_serializing && serde.skip_deserializing {
```
Is this really an error that we (Avro) should break the build ?
It is up to the user whether to optimize the serde attributes or not.

martin-g commented on avro_derive/tests/serde.rs:
```diff
@@ -0,0 +1,659 @@
+use apache_avro::{AvroSchema, Error, Reader, Schema, Writer, from_value};
+use serde::{Deserialize, Serialize, de::DeserializeOwned};
+
+/// Takes in a type that implements the right combination of traits and runs it through a Serde Cycle and asserts the result is the same
```
```suggestion
/// Takes in a type that implements the right combination of traits and runs it through a Serde round-trip and asserts the result is the same
```

martin-g commented on avro_derive/tests/serde.rs:
```diff
@@ -0,0 +1,659 @@
+    assert_eq!(obj, serde(obj.clone()).unwrap());
+}
+
+/// Takes in a type that implements the right combination of traits and runs it through a Serde Cycle and asserts that the error matches the expected string
```
```suggestion
/// Takes in a type that implements the right combination of traits and runs it through a Serde round-trip and asserts that the error matches the expected string
```

martin-g commented on avro_derive/src/attributes/mod.rs:
```diff
@@ -0,0 +1,206 @@
+            errors.push(syn::Error::new(
+                span,
+                "AvroSchema derive does not support different rename rules for serializing and deserializing (`rename_all(serialize = \"..\", deserialize = \"..\")`)"
+            ))
```
```suggestion
            ));
```

Kriskras99 commented:
I've fixed all the feedback. I've also deprecated `#[avro(name = "..")]` in favor of `#[serde(rename = "..")]` and added deprecation warnings on nightly.

martin-g commented on avro_derive/src/lib.rs:
This function could be moved to attributes/mod.rs
